File: arrays/SinglyLinkedList.java
package arrays;

public class SinglyLinkedList<T> {
    // Node
    private static class Node<T> {
        private T element;
        private Node<T> next;

        public Node(T e, Node<T> n) {
            element = e;
            next = n;
        }

        public T getElement() {
            return element;
        }

        public Node<T> getNext() {
            return next;
        }

        public void setNext(Node<T> n) {
            next = n;
        }
    }

    private Node<T> head = null;
    private Node<T> tail = null;
    private int size = 0;

    public SinglyLinkedList() {
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    public T first() {
        if (isEmpty()) {
            return null;
        } else {
            return head.getElement();
        }
    }

    public T last() {
        if (isEmpty()) {
            return null;
        } else {
            return tail.getElement();
        }
    }
}

File: linkedlists/singly_linked_list.dart
class Node<T> {
  T _element;
  Node<T>? _next;

  Node(this._element, this._next);

  T getElement() {
    return _element;
  }

  Node<T>? getNext() {
    return _next;
  }

  void setNext(Node<T>? n) {
    _next = n;
  }
}

class SinglyLinkedList<T> extends Node<T> {
  SinglyLinkedList(T element, Node<T> next) : super(element, next);

  int size = 0;
  Node<T>? head, tail = null;

  int getSize() {
    return size;
  }

  // check is size of list is empty
  bool isEmpty() {
    return size == 0;
  }

  T? first() {
    if (isEmpty()) return null;
    return head!.getElement();
  }

  T? last() {
    if (isEmpty()) return null;
    return tail!.getElement();
  }

  void addFirst(T element) {
    Node<T> newHead = Node<T>(element, head);
    if (isEmpty()) {
      tail = head;
    } else {
      newHead.setNext(head);
      head = newHead;
    }
    size++;
  }

  void addLast(T element) {
    Node<T> newTail = new Node<T>(element, null);
    if (isEmpty()) {
      head = newTail;
    } else {
      tail!.setNext(newTail);
    }
    tail = newTail;
    size += 1;
  }

  T? removeFirst() {
    if (isEmpty()) {
      return null;
    }
    T nextHead = head!.getElement();
    head = head!.getNext();
    size -= 1;
    if (isEmpty()) {
      tail = null;
    }
    return nextHead;
  }
}
